include/secp256k1_musig.h

```diff
@@ -0,0 +1,428 @@
+ *              msg: The 32-byte message (hash) to be signed
+ *       msg_is_set: Whether the above message has been set
+ *  has_secret_data: Whether this session object has a signers' secret data; if this
+ *                   is `false`, it may still be used for verification purposes.
+ *           seckey: If `has_secret_data`, the signer's secret key
+ *         secnonce: If `has_secret_data`, the signer's secret nonce
+ *            nonce: If `has_secret_data`, the signer's public nonce
+ *        nonce_commitments_hash: If `has_secret_data` and
+ *                   `nonce_commitments_hash_is_set`, the hash of all signers'
+ *                   commitments
+ * nonce_commitments_hash_is_set: If `has_secret_data`, whether the
+ *                   nonce_commitments_hash has been set
+ */
+typedef struct {
+    secp256k1_pubkey combined_pk;
+    size_t n_signers;
+    unsigned char pk_hash[32];
+    secp256k1_pubkey combined_nonce;
+    int nonce_is_set;
+    int nonce_is_negated;
+    unsigned char msg[32];
+    int msg_is_set;
+    int has_secret_data;
+    unsigned char seckey[32];
+    unsigned char secnonce[32];
+    secp256k1_pubkey nonce;
+    unsigned char nonce_commitments_hash[32];
+    int nonce_commitments_hash_is_set;
+} secp256k1_musig_session;
+
+/** Data structure containing data on all signers in a single session.
+ *
+ * The workflow for this structure is as follows:
+ *
+ * 1. This structure is initialized with `musig_session_initialize` or
+ *    `musig_session_initialize_verifier`, which set the `pubkey` and `index` fields,
+ *    and zero out all other fields. The public session is initialized with the
+ *    signers' nonce_commitments.
+ *
+ * 2. In a non-public session the nonce_commitments are set with the function
+ *    `musig_get_public_nonce`, which as a also returns the signer's public nonce.
+ *    This ensures that the public nonce is not exposed until all commitments have
+ *    been received.
+ *
+ * 3. Each individual data struct should be updated with `musig_set_nonce` once a
+ *    nonce is available. This function takes a single signer data struct rather than
+ *    an array because it may fail in the case that the provided nonce does not match
+ *    the commitment. In this case, it is desirable to identify the exact party whose
+ *    nonce was inconsistent.
+ *
+ * Fields:
+ *   present: indicates whether the signers nonce is set
+ *     index: index of the signer in the MuSig key aggregation
+ *    pubkey: public key that the signer will use for partial signing
+ *     nonce: public nonce, must be a valid curvepoint if the signer is `present`
+ * nonce_commitment: commitment to the nonce, or all-bits zero if a commitment
+ *                   has not yet been set
+ */
+typedef struct {
+    int present;
+    size_t index;
+    secp256k1_pubkey pubkey;
+    secp256k1_pubkey nonce;
+    unsigned char nonce_commitment[32];
+} secp256k1_musig_session_signer_data;
```

Comment:
How about not giving this to every function but instead only to the init functions. I think otherwise one way to misuse the current API is to use a session with the wrong signers' data.

Reply:
I agree that it's a problem that callers may use the session with the wrong signers. I hope I've caught and tested all cases where this could be a security issue.

I can't really imagine how the session would reasonably manage the signers though. The session can't allocate the signers on the stack because we don't have variable length arrays and we don't want to use them. Keeping a pointer in the session to signer data which is not manage by the session itself seems dangerous.

Reply:
Hm, yes, I can't find a case where this could a security problem. For the commitments, we have the commitments_hash. And for the public keys, the  combined_pk is set in stone already during initialization.

A little bit strange is that the pubkeys in the signer_data are only used for partial verification, and the init functions get the pubkeys and the combined_pk. If the pubkeys are anyway used only for partial verification, couldn't we drop them from init? (Or alternatively let init compute the combined_pk, at least in a singing session. Just for verification, the user can still use the combine function.)

Reply:
If you agree that moving pubkeys to `partial_verify` makes sense, then we could at least get rid of them in `session_initialize`.

The more I think about it, the more it makes sense to me. Maybe I'm oversimplifying but doesn't this even mean that you don't need to store them at all (if you don't care about partial verification)? I mean if the two of us just want to spend these coins under our combined public key pk, then we neither need to remember each other's public key nor send our public keys over the wire again. We just need to know H(pk0,pk1) and agree on our indices, but these can be canonical such as 0=software wallet, 1=hardware wallet. So it can even make sense to run the protocol without even knowing the other public keys (if I'm not mistaken, which could very well be).

Reply:
I moved pubkeys to partial verify